File: src/main/Java/boj_step16/boj9012.java
import java.io.*;
import java.util.Stack;

public class boj9012 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine());
        while (n-- > 0) writer.append((isVpn(reader.readLine()) ? "YES\n" : "NO\n"));
        writer.close();
    }
    private static boolean isVpn (String s) {
        Stack<Character> vpn = new Stack<>();
        for (char c: s.toCharArray()) {
            switch (c) {
                case '(': vpn.push('('); break;
                case ')': {
                    if (vpn.isEmpty()) return false;
                    else vpn.pop();
                } break;
            }
        }
        if (vpn.isEmpty()) return true;
        else return false;
    }
}

File: src/main/Kotlin/boj_step16/boj9012.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    val stack = Stack<Char>()
    repeat(n) {
        stack.clear()
        val bracket = r.readLine()
        bracket.forEach {
            if (it == '(') {
                stack.add(it)
            } else {
                if (stack.isEmpty()) {
                    w.write("NO\n")
                    return@repeat
                } else stack.pop()
            }
        }
        if (stack.isEmpty()) w.write("YES\n")
        else w.write("NO\n")
    }
    w.close()
}